### PR TITLE
fix: apiVersion and kind from CRD in registration

### PR DESCRIPTION
**Pull request by zregvart**
In `CustomResourceOperationsImpl` construction
`KubernetesDeserializer::registerCustomKind` is invoked without
`apiVersion` and with `kind` set to the class name. This could lead to
issues when deserializing custom resources, in my example:

```
com.fasterxml.jackson.databind.JsonMappingException: No resource type found for:camel.apache.org/v1alpha1#Integration
 at [Source: (String)"{"type":"DELETED","object":{"apiVersion":"camel.apache.org/v1alpha1","kind":"Integration","metadata":{"annotations":{"prometheus.io/port":"9779","prometheus.io/scrape":"true","syndesis.io/deploy-id":"i-M0SZstrdU_fCuujYkqGz:2","syndesis.io/deployment-version":"2","syndesis.io/integration-id":"i-M0SZstrdU_fCuujYkqGz","syndesis.io/integration-name":"t2l"},"creationTimestamp":"2020-02-19T13:14:10Z","generation":1,"labels":{"syndesis.io/app":"syndesis","syndesis.io/component":"integration","syndesis."[truncated 3644 chars]; line: 1, column: 4142] (through reference chain: io.fabric8.kubernetes.api.model.WatchEvent["object"])
	at com.fasterxml.jackson.databind.JsonMappingException.from(JsonMappingException.java:271) ~[jackson-databind-2.10.2.jar:2.10.2]
	at io.fabric8.kubernetes.internal.KubernetesDeserializer.deserialize(KubernetesDeserializer.java:78) ~[kubernetes-model-4.6.1.jar:4.6.1]
	at io.fabric8.kubernetes.internal.KubernetesDeserializer.deserialize(KubernetesDeserializer.java:33) ~[kubernetes-model-4.6.1.jar:4.6.1]
```

Because only `"Integration"` would be registered, whereas type lookup is
using `apiVersion` and `kind.

This changes that to specify the `apiVersion` and `kind` based on the
definition given in the `CustomResource`.

### CODE REVIEW

**Comment by rohanKanojia** [comment]
ok to test

File: kubernetes-client/src/test/java/io/fabric8/kubernetes/client/dsl/internal/CustomResourceOperationsImplTest.java [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+		JsonFactory factory = new MappingJsonFactory();
+		JsonParser parser = factory.createParser("{\n" + 
+		    "    \"apiVersion\": \"custom.group/v1alpha1\",\n" + 
+		    "    \"kind\": \"MyCustomResource\"\n" + 
+		    "}");
+
+		KubernetesDeserializer deserializer = new KubernetesDeserializer();
+		KubernetesResource resource = deserializer.deserialize(parser, null);
+
+		assertThat(resource, instanceOf(MyCustomResource.class));
```
Comment (manusa) [inline kubernetes-client/src/test/java/io/fabric8/kubernetes/client/dsl/internal/CustomResourceOperationsImplTest.java]: I'd probably add an assertion validation the content of a field in order to verify deserialization worked.

(e.g. `assertThat(((MyCustomResource)resource).getApiVersion(), equalTo("custom.group/v1alpha1"))`)

File: kubernetes-client/src/test/java/io/fabric8/kubernetes/client/dsl/internal/CustomResourceOperationsImplTest.java [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+
+public class CustomResourceOperationsImplTest {
+
+	public static class MyCustomResource extends CustomResource {
+    }
+
+	public static class MyCustomResourceList extends CustomResourceList<MyCustomResource> {
+    }
+
+    CustomResourceDefinition crd = new CustomResourceDefinitionBuilder()
```
Comment (manusa) [inline kubernetes-client/src/test/java/io/fabric8/kubernetes/client/dsl/internal/CustomResourceOperationsImplTest.java]: Please add access modifiers to these field (private)

Ideally field instantiation should happen inside a test setup method (`@BeforeEach`) in order to avoid side effects when new tests are added to the suite.

File: kubernetes-client/src/test/java/io/fabric8/kubernetes/client/dsl/internal/CustomResourceOperationsImplTest.java (hunk 1 above)
Comment (zregvart) [reply kubernetes-client/src/test/java/io/fabric8/kubernetes/client/dsl/internal/CustomResourceOperationsImplTest.java]: Added https://github.com/fabric8io/kubernetes-client/pull/2011/files#diff-1500b8672a6f6391abfa76c29b8128cfR83

File: kubernetes-client/src/test/java/io/fabric8/kubernetes/client/dsl/internal/CustomResourceOperationsImplTest.java (hunk 2 above)
Comment (zregvart) [reply kubernetes-client/src/test/java/io/fabric8/kubernetes/client/dsl/internal/CustomResourceOperationsImplTest.java]: Changed to `private final` this should cause no side effects that I can think of.

**Comment by zregvart** [comment]
@manusa thanks, can you have another look?

File: kubernetes-client/src/test/java/io/fabric8/kubernetes/client/dsl/internal/CustomResourceOperationsImplTest.java (hunk 2 above)
Comment (manusa) [reply kubernetes-client/src/test/java/io/fabric8/kubernetes/client/dsl/internal/CustomResourceOperationsImplTest.java]: Making it final doesn't make it immutable which may lead to some side-effects:

So if one of the test cases (not at the moment, and probably not ever) does something like `crd.setApiVersion("bar")` in order to do some corner-case test and some other test asserts on this field, it will fail. It will also fail in a non-deterministic way depending on how the system computes the test execution order.

That's why it's always recommended to use setup and tear down methods in tests.

**Comment by rohanKanojia** [comment]
[merge]

**Comment by rohanKanojia** [comment]
[merge]

**Comment by rohanKanojia** [comment]
[merge]
